src/wfcomman.c

```diff
@@ -1632,28 +1632,8 @@ AppCommandProc(register DWORD id)
 
    case IDM_FORMAT:
 
-      if (CancelInfo.hCancelDlg) {
-         SetFocus(CancelInfo.hCancelDlg);
-         break;
-      }
-
-      if (CancelInfo.hThread) {
-         //
-         // Don't create any new worker threads
-         // Just create old dialog
-         //
-
-         CreateDialog(hAppInstance, (LPTSTR) MAKEINTRESOURCE(CANCELDLG), hwndFrame, (DLGPROC) CancelDlgProc);
-
-         return TRUE;
-      }
-
-      if (!FmifsLoaded())
-         break;
-
-      // Don't use modal dialog box
-
-      FormatDiskette(hwndFrame,FALSE);
+      // Don't use modal dialog box, set hWndParent = NULL
+      DialogBox(hAppInstance, (LPTSTR)MAKEINTRESOURCE(FORMATSELECTDLG), NULL, (DLGPROC)FormatSelectDlgProc);

```

Comment:
In order for the dialog to be positioned like the others, we need to pass hwndFrame as the third parameter to DialogBox.

Reply:
@craigwi Good point, and I actually originally wrote the code that way. I later changed the parameter to NULL to avoid making this dialog modal. Really, I was more concerned about the dialog created by SHFormatDrive being modal, because formatting can be a long-running operation. SHFormatDrive requires a parent HWND (says MSDN), so I used the first dialog (the one you commented on) as its parent. Therefore, to prevent a situation where both dialogs were modal, I used NULL in the line you mentioned.

I see a few options:
1. Just change the line you called out and make hWndParent=hwndFrame. That will solve the positioning issue, but will make the dialogs modal.
2. Make your suggested change, and also change the call to SHFormatDrive so that it its parent HWND is NULL, to avoid the modal side effect. This goes against MSDN's guidance, so I'm not sure what the side effects may be, although it seems to work.
3. Make your suggested change, and also change the call to SHFormatDrive so that it its parent HWND is a hidden window. This avoids the modal problem while still complying with MSDN's recommendation. This is actually what explorer.exe seems to do, and it seems the proper thing to do, but I had avoided this approach before since involved extra window management.

What do you think?

Reply:
I suggest making the FORMATSELECTDLG modeless by calling CreateDialog instead of DialogBox.  You can also save the hwnd returned to activate it so there is only one.  I tried this out and the basic cases seem to work.

There are several other examples of modeless dialogs (e.g., CancelDlg) that you can use as a pattern.

I would like to see the FileCopy dialog becomes modelsss in this way.  It has always annoyed me that it was modal.

Reply:
Good suggestion; thanks! I had forgotten about CreateDialog. It has been a while since I've written C/Win32 code! I'll look into making that change.